### PR TITLE
fix(validation): Throw an exception when referencing a non existing env

### PR DESCRIPTION
**Pull request by gal-yardeni**
fixes https://github.com/spinnaker/keel/issues/936.

Adding a validation for the depends on constraint - if referencing to a non-existing environment, throw a bad request exception. 

I'm planning on extracting the delivery config validation into a new service but will do it in a separate PR.

### CODE REVIEW

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt [hunk 1]
```diff
@@ -231,6 +233,21 @@ class CombinedRepository(
       log.error("Validation failed for ${config.name}, duplicate artifact references found: $duplicatesArtifactNameToRef")
       throw DuplicateArtifactReferenceException(duplicatesArtifactNameToRef)
     }
+
+    /**
+     * validating depends on environments
+     */
+    config.environments.forEach { environment ->
+      environment.constraints.forEach { constraint ->
+        when (constraint is DependsOnConstraint) {
```
Comment (emjburns) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt:242]: I'd just use an if statement here

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/MissingEnvironmentReferenceException.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.netflix.spinnaker.keel.exceptions
+
+class MissingEnvironmentReferenceException(
+  private val environmentName: String
+) : ValidationException(
+  "The depend-on environment named: $environmentName " +
+    "is missing in the config. Please ensure you reference to the right environment."
```
Comment (emjburns) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/MissingEnvironmentReferenceException.kt:7]:
```suggestion
    "is missing in the config. Please ensure you reference the right environment."
```

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt [hunk 3]
```diff
@@ -408,6 +410,48 @@ abstract class CombinedRepositoryTests<D : DeliveryConfigRepository, R : Resourc
...
+        val submittedConfig = SubmittedDeliveryConfig(
+          name = configName,
+          application = "keel",
+          serviceAccount = "keel@spinnaker",
+          artifacts = setOf(DockerArtifact(name = "org/thing-1", deliveryConfigName = configName, reference = "thing")),
+          environments = setOf(
+            SubmittedEnvironment(
+              name = "test",
+              resources = setOf(
+                SubmittedResource(
```
Comment (emjburns) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt:424]: you can probably make these just empty sets (environments with no resources) - I think we allow that, right?

Comment (emjburns) [reply keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt:424]: (makes the test easier to understand and maintain

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt [hunk 4]
```diff
@@ -408,6 +410,48 @@ abstract class CombinedRepositoryTests<D : DeliveryConfigRepository, R : Resourc
...
+          )
+        )
+
+        test("an error is thrown and config is not persisted") {
+          expectCatching {
+            subject.upsertDeliveryConfig(submittedConfig)
+          }.failed()
+            .isA<MissingEnvironmentReferenceException>()
+
+          expectThat(subject.allResourceNames().size).isEqualTo(0)
```
Comment (emjburns) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt:452]: if you remove the resources like suggested above you can just check to make sure the delivery config is not persisted.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt (hunk 1 above)
Comment (gal-yardeni) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt:242]: good call, will change

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt (hunk 3 above)
Comment (gal-yardeni) [reply keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt:424]: done

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt (hunk 4 above)
Comment (gal-yardeni) [reply keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt:440]: I couldn't find the right statement for making sure the delivery config table is empty (I saw this check is used for the same thing, so I used it).. can you point me to the right statement?

Comment (emjburns) [reply keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt:440]: I would just do:
```
expectCatching {
            subject.getDeliveryConfig(name)
          }.failed()
            .isA<NoSuchDeliveryConfigException>()
```

Comment (gal-yardeni) [reply keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt]: oooo right! thank you!

**Pull request merged**